### ISSUE TITLE
Issue to track NSURLProtocol for WKWebView radars

### ISSUE THREAD

**Issue by garvankeeley** [brave/browser-ios#96]
All of these are the same,  asking for NSURLProtocol in WKWebView.

http://www.openradar.me/17190141
http://www.openradar.me/17240707
http://www.openradar.me/18425900
http://www.openradar.me/18492325
http://www.openradar.me/24231298
https://openradar.appspot.com/24231298

**Comment by garvankeeley** [comment]
Browsers that appear to use NSURLProtocol filtering:
* Chrome 
* Dolphin
• Puffin
* iCab
* Adblock Plus Browser
* Ghostery

Opera: need to check.

WKWebView: Firefox and Sleipnir

**Comment by pavel-zdenek** [comment]
I confirm Adblock Browser and (the underlying) Kitt being based on UIWebView/NSURLProtocol. We know that Opera uses UIWebView - which does not imply NSURLProtocol, sure. Our rule of thumb is the browser reaction to a memory pressure when running http://dromaeo.com/?all. UIWebView ones obviously crash completely. Regarding the issue in general: we even spent one precious Apple TSI a year ago to ask for having it fixed. Apple does not consider it a bug so resolutely that they credited back the TSI.

**Comment by garvankeeley** [comment]
Hi, that is a good site for testing, thanks for the link.
I based the list on if the browser has network filtering that requires URL interception (Opera and Chrome with their data compression features, ad blocking in the other browsers) then they are using NSURLProtocol.

**Comment by pavel-zdenek** [comment]
Chrome as of yesterday's update 48.0.2564.87 has moved to WKWebView

**Comment by garvankeeley** [comment]
Chrome dropped features to go to WKWebView:
<img width="390" alt="screenshot 2016-01-28 10 31 20" src="https://cloud.githubusercontent.com/assets/5495083/12648949/63907508-c5aa-11e5-9db6-d562f89e7262.png">

**Comment by jswindler-ghostery** [comment]
Ghostery does indeed use NSURLProtocol / NSURLCache with a UIWebView. We'd love to be able to upgrade to WKWebView but that doesn't appear to be an option at this point.

**Comment by garvankeeley** [comment]
We keep hearing rumours that Apple is planning to add network request filtering API to WKWebView, the sooner the better for all of us 3rd party browser people. 
At a minimum something like (apologies for silly func naming):

`func urlForURL(webview: WKWebView, URL: NSURL) -> NSURL`
• would allow one to replace the URL before loading, the webview would then store this URL in history

`func requestForRequest(webview: WKWebView, request: NSURLRequest) -> NSURLRequest`
• would allow one to replace the network request with another request, redirecting it without the webview 'knowing' about the redirection

Even that one could do custom caching with, redirecting requests to an in-app web server (clunky, but could be done) .

**Comment by garvankeeley** [comment]
@diegodoval curious on your opinion: what do you think a minimal callback API would look like?

**Comment by jswindler-ghostery** [comment]
Something else that would be very helpful to have is some kind of delegate that gets called when a given request has completed. We want to be able to record how much time each request takes (specifically 3rd party tracker requests) as a tool for people to debug their mobile websites.

**Comment by diegodoval** [comment]
I think that a minimal API would allow 

a) processing/filtering the NSURLRequest before anything happens
b) getting called when the response code & headers are received
c) getting called when the body download completes/fails (if applicable)

a) is the most important for our purposes and would cover the cases you mention, doing b) and c) would also be extremely useful, even if it was only as an observer and able to do anything only after the fact (eg through JS directly on the webpage after it loads/executes)

**Comment by WildDylan** [comment]
Hey, i used [This](https://github.com/WildDylan/WKWebViewWithURLProtocol) for solve this question.

It's easy to let URLProtocol hook WKWebView with Runtime!

**Comment by pavel-zdenek** [comment]
@WildDylan sorry to disappoint you, but the API you're hooking on is private, which means that you can't use the technique in an AppStore-published app. See e.g. http://www.openradar.me/18492325

**Comment by WildDylan** [comment]
@pavel-zdenek  what do you think of `private method check in Apple's review`. 
First: `registerSchemeForCustomProtocol:` this method when we get with:

```objc
SEL sel = NSSelectorFromString(@"registerSchemeForCustomProtocol:");
```
will be ok.

Second: the `WKBrowsingContextController ` class. if we use:

```objc
Class cls = NSClassFromString(@"WKBrowsingContextController");
```
that's may be checked when in review. because Apple's review will check all class name. `WKBrowsingContextController` is private and easier to find.

so i use:

```objc
Class cls = [[[WKWebView new] valueForKey:@"browsingContextController"] class];
``` 
KVC to get this private class. it's safe.

And also, if you worry it, you can use `jsPatch`, dynamic do this.

**Comment by garvankeeley** [comment]
Cool! Trying it out now. As for passing app store, I have had lots of similar KVC code get into the app store (sometimes have to split up the strings used). But I have only ever used it to solve edge-case problems so that if the code needed to be pulled, the app would still be functional.

**Comment by pavel-zdenek** [comment]
@WildDylan @garvankeeley ok i admit dropping my weapons and running away whenever runtime introspection of private apis is mentioned, without knowing the intent obscuring subtleties. So does it really work dependably? The example code has empty startLoading
https://github.com/WildDylan/WKWebViewWithURLProtocol/blob/master/Example/WKWebViewWithURLProtocol/URLProtocol.m#L33
so it does not ensure the request completeness. There are rumors that it's missing HTTP POST body and possibly other stuff
https://bugs.webkit.org/show_bug.cgi?id=137299
https://bugs.webkit.org/show_bug.cgi?id=140188

**Comment by WildDylan** [comment]
@pavel-zdenek  Hi, i complete this demo project. webView loaded the demo html from URLProtocol.

**Comment by TianleZhang** [comment]
@WildDylan Thx for your solution. I did implemented as well. It works like a charm, however I also confirm the issue @pavel-zdenek mentioned https://bugs.webkit.org/show_bug.cgi?id=137299. The HTTPbody is nil. Do you have any ideal for that as well?

**Comment by WildDylan** [comment]
Hey, @TianleZhang, HTTPBody will be ignored in WKWebView, [bug](https://forums.developer.apple.com/thread/18952), resolve with `[Javascript bridge](http://stackoverflow.com/questions/26253133/cant-set-headers-on-my-wkwebview-post-request/26342224#26342224)` or `NSURLSession`.  Send POST request use NSURLSession, load html with wkwebView.

**Comment by gguerini** [comment]
Hey @WildDylan, your solution solved most of my problems. I was able to get my `WKWebViews` persist the headers like I used to do with `UIWebViews` + custom `NSURLProtocol`. Thank you so much for that!!!

But now I'm facing the empty body issue. The JS solution you provided it's too cumbersome because my application has many different forms and it would be really hard to manually set the attributes for each form ([as suggested here](http://stackoverflow.com/questions/26253133/cant-set-headers-on-my-wkwebview-post-request/26342224#26342224)).

You mentioned another solution: `NSURLSession`. Could you please elaborate more? I have no idea how to even start.

Thank you so much.

**Comment by pavel-zdenek** [comment]
@gguerini he means doing the request yourself in native code instead of asking WebKit through JS, and then injecting the result back to WKWebView. There is enough examples on Apple dev site on how to use `NSURLSession` for `POST` requests
https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/URLLoadingSystem/Articles/UsingNSURLSession.html#//apple_ref/doc/uid/TP40013509-SW15

But if you are having some issue with coding JS support for many different forms, NSURLSession would require you to do even more coding for the same number of forms. You basically need to construct the same requests as the forms do, in native code.

**Comment by garvankeeley** [comment]
I have a branch `brave-wkwebview-proto` using NSURLProtocol on WKWebView, and also did a test upload to the store and it passed. Of course, the post body removal blocks any further progress. 

The URL request header will specify the request is a POST, but checking this and returning false from canInitWithRequest doesn't workaround the bug.

https://bugs.webkit.org/show_bug.cgi?id=138169

**Comment by TianleZhang** [comment]
@gguerini I tried both ways suggest by @WildDylan. Unfortunately both does not work. In the first way, you do JS bridging, the NSURLProtocol will intercept the request and your body is gone after the intercept. So it is becoming a chicken egg problem. And I tried the session way as well by putting the body in the header during intercept read it from header and set it to the body. However, I got a bad memory access error during the request call. I think this may have to do with the actual body memory is in the process of WKWebview so it can not read the memory location from the app or something like that. 

In the end, we manage to solve the problem in a very crazy way. We over written the XHR obj in javascript and route all the network traffic to native and handle it with AFNetworking. So basically saying, we had our own native proxy server does all the traffic and only use WKWebview for rendering, in this case, all the CORS issues are gone, and we can intercept & modify the request and response in the way we want.

**Comment by WildDylan** [comment]
@TianleZhang Awesome what you have done! But some times, we use NSURLProtocol for WKWebView  is want to handle the resources loading or caching. Such as <img />, .css, .js, resources can be managed in native. And about HTTP body, NSURLProtocol always ignored body where request load from WKWebView, The Scene is:`[_webView loadRequest: yourRequest]`, if load `yourRequest` with UIWebView, body of `yourRequest ` can be hooked by NSURLProtocol, so i suggest, use header field (  header fields can be hooked in Protocol, we have some test about it), when you get fields in URLProtocol, use NSURLSession or what you want for get response, and use client of URLProtocol process the response.

**Comment by TianleZhang** [comment]
@WildDylan Yes, I agree what you are saying, luckily that was not my part of requirement yet. And for intercepting the caching or resources loading, we don't need to do a POST with body.  So I think if we want to have a complete intercept, before Apple fix the issue, we should implement both approaches.

**Comment by firuz1844** [comment]
@TianleZhang can you please share you realisation of XHR object in JavaScript? We are trying to do same thing but haven't a mind where to dig.

**Comment by TianleZhang** [comment]
@firuz1844 I can't really share the code with you. But this is how it how we implemented. We basically replace the XMLHttpRequest, and in the send function, we call the js to call to our own network API 

window.XMLHttpRequest = function() {
        'use strict';
        var t = this,
            _method,
            _url,
            _origin,
            _requestHeaders = {},
            _responseHeaders = {},
            _listeners = [],
            _nop = function() { return; };

        t.withCredentials = true;

        t.open = function (method, url) {
            console.log(url);
            _method = method;
            _url = url;
            var a = document.createElement('a');
            a.href = url;
            _origin = a.origin;
        };

        t.setRequestHeader = function(header, value) {
            _requestHeaders[header] = value;
        };

        t.getResponseHeader = function(header) {
            return _responseHeaders[header];
        };

        t.getAllResponseHeaders = function() {
            return _responseHeaders;
        };

        t.addEventListener = function(name, callback) {
            _listeners.push( { name: name,  callback: callback } );
        };

       t.send = function() {
               call.my.network.function.from native();
        }
...

**Comment by firuz1844** [comment]
@TianleZhang thanks for you answer! Is it possible to get body of request with your approach? Where it better to catch on open() or on send()?

**Comment by TianleZhang** [comment]
@firuz1844 Yes. with this approach, you are possible to catch the entire requests and response. From the send(), you call to your native HttpSession (in our case we used AFNetworking.) and you can intercept the request and response in native.

**Comment by WATER1350** [comment]
@TianleZhang how do you return response to WKWebView after you hook ajax requests and reload them by AFNetworking.

**Comment by firuz1844** [comment]
@WATER1350 I am saving body value in dictionary and retrieve it from native code by evaluating JavaScript. After that I am removing body in "onreadystatechange" event if state == XMLHttpRequest.DONE

```
var _proxysend = window.XMLHttpRequest.prototype.send;
window.XMLHttpRequest.prototype.send = function() {
 // Saving body
        if (arguments[0]) {
            if (window.body_dictionary == undefined) {
                window.body_dictionary = new Object();
            }
            window.body_dictionary[this._url.hashCode()] = arguments[0].toString();
        }
     return _proxysend.apply(this, arguments);
}
```

**Comment by microyulive** [comment]
@WildDylan  Hello! Is it possible to check url like "www.google.com" but not scheme like "http"?
